Exchange co2signal package with aioelectricitymaps

jpbede opened the pull request:
## Proposed change
The package used for co2signal integration seems to have been abandoned for 5 years. I did the work and wrote a new async package.

The new package uses the new API endpoint by default. When the package detects an old key, it uses the old API, but it doesn't seem to have authentication or rate limiting anymore. I've tested this with an old and new API key. 

https://github.com/jpbede/aioelectricitymaps

 A side effect of this is smaller and cleaner coordinator code.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jpbede committed:
Exchange co2signal package with aioelectricitymaps

jpbede committed:
Bump aioelectricitymaps to 0.1.3

bdraco commented on homeassistant/components/co2signal/config_flow.py:
```diff
@@ -117,19 +117,26 @@ async def _validate_and_create(
-                title=get_extra_name(data) or "CO2 Signal",
-                data=data,
-            )
+        async with ElectricityMaps(token=data[CONF_API_KEY]) as em:
```
Can you avoid creating a new aiohttp session here and use the helper ?

https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=aiohttp#platinum-
> Uses aiohttp or httpx and allows passing in websession (if making HTTP requests)

jpbede replied:
Yea, prepared the package for this, but totally forgotten to add it here 🤦

bdraco commented on tests/components/co2signal/test_diagnostics.py:
```diff
@@ -27,7 +27,7 @@ async def test_entry_diagnostics(
     )
     config_entry.add_to_hass(hass)
-    with patch("CO2Signal.get_latest", return_value=VALID_PAYLOAD):
+    with patch("aioelectricitymaps.ElectricityMaps._get", return_value=VALID_PAYLOAD):
```
Ideally we patch the library where its imported instead of globally

bdraco commented on tests/components/co2signal/test_config_flow.py:
```diff
@@ -180,7 +173,7 @@ async def test_form_error_handling(hass: HomeAssistant, side_effect, err_code) -
 
     with patch(
-        "CO2Signal.get_latest",
+        "aioelectricitymaps.ElectricityMaps._get",
```
Ideally we patch the library where its imported instead of globally

bdraco commented on tests/components/co2signal/test_config_flow.py:
```diff
@@ -165,7 +158,7 @@ async def test_form_error_handling(hass: HomeAssistant, side_effect, err_code) -
 
     with patch(
-        "CO2Signal.get_latest",
+        "aioelectricitymaps.ElectricityMaps._get",
```
Ideally we patch the library where its imported instead of globally

bdraco commented on tests/components/co2signal/test_config_flow.py:
```diff
@@ -22,7 +26,7 @@ async def test_form_home(hass: HomeAssistant) -> None:
 
     with patch(
-        "CO2Signal.get_latest",
+        "aioelectricitymaps.ElectricityMaps._get",
```
Ideally we patch the library where its imported instead of globally

bdraco commented on tests/components/co2signal/test_config_flow.py:
```diff
@@ -64,7 +68,7 @@ async def test_form_coordinates(hass: HomeAssistant) -> None:
 
     with patch(
-        "CO2Signal.get_latest",
+        "aioelectricitymaps.ElectricityMaps._get",
```
Ideally we patch the library where its imported instead of globally

bdraco commented on tests/components/co2signal/test_config_flow.py:
```diff
@@ -108,7 +112,7 @@ async def test_form_country(hass: HomeAssistant) -> None:
 
     with patch(
-        "CO2Signal.get_latest",
+        "aioelectricitymaps.ElectricityMaps._get",
```
Ideally we patch the library where its imported instead of globally

bdraco commented on homeassistant/components/co2signal/__init__.py:
```diff
@@ -13,7 +15,9 @@
     """Set up CO2 Signal from a config entry."""
-    coordinator = CO2SignalCoordinator(hass, entry)
+    coordinator = CO2SignalCoordinator(
+        hass, ElectricityMaps(token=entry.data[CONF_API_KEY])
```
Can you avoid creating a new aiohttp session here and use the helper ?

https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=aiohttp#platinum-
> Uses aiohttp or httpx and allows passing in websession (if making HTTP requests)

joostlek commented on tests/components/co2signal/test_config_flow.py:
```diff
@@ -22,7 +26,7 @@ async def test_form_home(hass: HomeAssistant) -> None:
 
     with patch(
-        "CO2Signal.get_latest",
+        "aioelectricitymaps.ElectricityMaps._get",
```
Can we patch the public function we are calling? And I know this didnt happen before, but can we patch it at the place where its used? `homeassistant.components.co2signal.config_flow.ElectricityMaps.....`

jpbede replied:
Yes, this should be possible, but I wanted to keep the tests as original as possible for this PR. I can change this in a later PR. 

The "patch place" will change in a second... or two.... :)

joostlek replied:
haha great

joostlek commented on homeassistant/components/co2signal/sensor.py:
```diff
@@ -24,7 +26,19 @@
+class ElectricityMapsMixin:
+    """Mixin for value and unit_of_measurement_fn function."""
+
+    value_fn: Callable[[CarbonIntensityResponse], float | None] = field(
+        default=lambda response: None
+    )
```
```suggestion
    value_fn: Callable[[CarbonIntensityResponse], float | None]
```

joostlek commented on homeassistant/components/co2signal/sensor.py:
```diff
@@ -24,7 +26,19 @@
+    value_fn: Callable[[CarbonIntensityResponse], float | None] = field(
+        default=lambda response: None
+    )
+    unit_of_measurement_fn: Callable[[CarbonIntensityResponse], str | None] = field(
```
I think you can make this one `Callable[] | None` and put it in `CO2SensorEntityDescription` and default to `None`

jpbede replied:
Hm that will not work, `None` is not callable, if I got it right :)

jpbede committed:
Implement suggestions from code review

joostlek commented on homeassistant/components/co2signal/sensor.py:
```diff
@@ -29,20 +29,18 @@
 
     # For backwards compat, allow description to override unique ID key to use
     unique_id: str | None = None
+    unit_of_measurement_fn: Callable[[CarbonIntensityResponse], str | None] = field(
+        default=lambda response: None
+    )
```
```suggestion
    unit_of_measurement_fn: Callable[[CarbonIntensityResponse], str | None] | None = None
```

jpbede replied:
Okay, I changed the function `native_unit_of_measurement` for this then :)

jpbede committed:
Change patch in diagnostics test

jpbede committed:
Change native_unit_of_measurement function

joostlek commented:
I would suggest changing
https://github.com/home-assistant/core/blob/6f845497846c3463d151124555921d9d14b3cc11/homeassistant/components/co2signal/diagnostics.py#L25

`"data": asdict(coordinator.data},`

jpbede commented:
Makes sense

jpbede committed:
Implement some more suggestions

bdraco commented:
I tried this PR on my production and got

<img width="655" alt="Screenshot 2023-10-20 at 9 55 59 AM" src="https://github.com/home-assistant/core/assets/663432/191598c8-501a-4e8a-adfa-d240586472ac">

bdraco commented:
Let me see if there is anything useful in the log

bdraco commented:
```
2023-10-20 14:55:17.882 WARNING (MainThread) [homeassistant.config_entries] Config entry 'CO2 Signal' for co2signal integration could not authenticate
2023-10-20 14:55:17.929 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 861, in async_init
    flow, result = await task
                   ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 889, in _async_init
    result = await self._async_handle_step(flow, flow.init_step, data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 385, in _async_handle_step
    self._raise_if_step_does_not_exist(flow, step_id)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 440, in _raise_if_step_does_not_exist
    raise UnknownStep(
homeassistant.data_entry_flow.UnknownStep: Handler ConfigEntriesFlowManager doesn't support step reauth
```

joostlek commented on homeassistant/components/co2signal/coordinator.py:
```diff
@@ -1,94 +1,60 @@
+                    ),
+                )
+            except InvalidToken as err:
+                raise ConfigEntryAuthFailed from err
```
```suggestion
                raise ConfigEntryError from err
```
This would make the integration back down iirc

bdraco replied:
This is an existing bug, the integration raises `ConfigEntryAuthFailed` but `async_step_reauth` is not implemented

bdraco commented on homeassistant/components/co2signal/manifest.json:
```diff
@@ -7,5 +7,5 @@
   "integration_type": "service",
   "iot_class": "cloud_polling",
   "loggers": ["CO2Signal"],
```
Looks like the updating the loggers was missed

bdraco commented:
I tried turning on debug logging for `aioelectricitymaps` but nothing more in the logs

jpbede commented:
Yes, I think detecting whether the old API is being used or not is not ideal at this time. Will change this in the package tomorrow.

jpbede commented:
@bdraco just a quick question: are you using an old API key, from the old API?

bdraco commented:
> @bdraco just a quick question: are you using an old API key, from the old API?

I only set up the integration for testing a while ago so I honestly don't know if its from the old or new API.

Its 32 characters long in case that helpers

jpbede commented:
Okay, thanks

jpbede commented:
Hm, based on the character length I would guess it is a token for the new one (since my token is also 32 characters long), old tokens are shorter AFAIK.
Could you compare your token with the one on https://api-portal.electricitymaps.com/keys?

The integration worked with this token so far, since the old API doesn't require a token anymore, so you could enter anything in the config flow :)

To me it feels more and more like we should migrate completely to the new API, but that would be a breaking change for those who have an old token.

bdraco commented:
I'm happy to share the API key with you, if you send me a message on discord, same handle

jpbede committed:
Bump aioelectricitymaps and some minor fixes

jpbede committed:
Bump aioelectricitymaps to 0.1.5 and avoid update on adding sensors

jpbede commented:
I added a retry mechanism to the package that first tries the new API, and if an invalid token is detected, the package reverts to the legacy API. Logging has also been added to the package.

In addition, I fixed small bug  that causes the API to pull twice when adding sensors. 
Also the update coordinator won't raise a `ConfigEntryAuthFailed` anymore. Will add this in another PR.

jpbede committed:
Add coordinator.py to .coveragerc

bdraco commented on homeassistant/components/co2signal/config_flow.py:
```diff
@@ -117,19 +118,27 @@ async def _validate_and_create(
+        session = async_get_clientsession(self.hass)
+        async with ElectricityMaps(token=data[CONF_API_KEY], session=session) as em:
+            try:
+                if CONF_COUNTRY_CODE in data:
+                    await em.latest_carbon_intensity_by_country_code(
+                        code=data[CONF_COUNTRY_CODE]
+                    )
+                else:
+                    await em.latest_carbon_intensity_by_coordinates(
+                        lat=data.get(CONF_LATITUDE, self.hass.config.latitude),
+                        lon=data.get(CONF_LONGITUDE, self.hass.config.longitude),
+                    )
```
This looks similar to 
https://github.com/home-assistant/core/pull/101955/files#diff-e751aa6f65cfe751a820fef40f11ba70c0047184a78641055790590a5b4c17d3R44

I think you could extract the inner block into a helper function to make it more dry

jpbede replied:
Addressed, moved to a helper function

bdraco commented:
Tested, everything working great @jpbede 

Thanks!

One small review comment above

I learned a new UOM today https://lowcarbonpower.org/blog/emissions

joostlek reviewed (approved):
LGTM once bdraco's comments are addressed

jpbede committed:
Move duplicate code to helper function

jpbede commented:
> I learned a new UOM today https://lowcarbonpower.org/blog/emissions

Oh, thanks for the article, I didn't know much about the UOM also

bdraco commented:
Thanks @jpbede

jpbede commented:
Thanks @bdraco @joostlek
